slash_commands/adorasac.js

```diff
@@ -0,0 +1,134 @@
+TOWER_PRIORITIES.set('heli_pilot', -500); // can't be placed on midnight mansion
+TOWER_PRIORITIES.set('monkey_ace', -380); // weird in 2mp
+TOWER_PRIORITIES.set('sniper_monkey', -250); // weird in 2mp
+TOWER_PRIORITIES.set('druid_monkey', -120); // jungle druid is weird in 2mp
+
+builder = new SlashCommandBuilder()
+    .setName('adorasac')
+    .setDescription('Calculate the optimal ewwww adora sacrifice to advance a certain amount of XP.')
+    .addIntegerOption((option) => option.setName('xp').setDescription('target XP to gain').setRequired(true).setMinValue(0).setMaxValue(XP_CAP))
+    .addStringOption((option) => option.setName('excluded_towers').setDescription('Comma-separated list of towers to exclude'));
+
+function parseExcludedTowers(excludedTowers) {
+    if (!excludedTowers) {
+        return []
+    }
+    let result = []
+    let excludedTowersList = excludedTowers.split(/\s*,\s*/);
+    for (let excludedTower of excludedTowersList) {
+        if (isTower(excludedTower)) {
+            throw new RangeError(`${excludedTower} is an invalid tower`);
+        }
+        result.push(Aliases.getCanonicalForm(excludedTower))
+    }
+    return result
+}
+
+async function execute(interaction) {
+    try {
+        let excludedTowers = parseExcludedTowers(interaction.options.getString('excluded_towers'));
+
+        let xp = interaction.options.getInteger('xp');
+
+        if (xp < 0 || xp > XP_CAP) {
+            throw new RangeError(`Provided XP value ${xp} should be between 0 and ${XP_CAP} inclusive`);
+        }
+
+        // optimization: use money divided by 5 to shrink DP space as all tower/upgrade costs are divisble by 5
+        let moneyToSpendDiv5 = Math.ceil(xp / (4 * 5));
+
+        // set up DP arrays
+        let bagSize = new Array(XP_CAP_ADJUSTED+1);
+        bagSize.fill(INF_BAG_SZ);
+        bagSize[0] = 0
+        let lastItem = new Array(XP_CAP_ADJUSTED+1);
+        lastItem.fill(0);
+
+        const costsDiv5ToUpgrade = new Map();
+
+        let towers = allTowers()
+        // sort towers in descending priority order, high priority should be looked at first
+        towers.sort((a, b) => (TOWER_PRIORITIES.get(b) || 0) - (TOWER_PRIORITIES.get(a) || 0));
+        for (let tower of towers) {
+            if (excludedTowers.indexOf(tower) < 0) {
+                for (let xpathSet of allUpgradeCrosspathSets()) {
+                    // exclude t5 beast handlers due to merge requirements
+                    // exclude monkeyopolis
+                    if (
+                        (tower !== 'beast_handler' || xpathSet.indexOf('5') < 0)
+                        && (tower !== 'monkey_village' || xpathSet[2] !== '5')
+                    ) {
+                        let processedCost = Math.ceil(costOfTowerUpgradeSet(tower, xpathSet, 'hard') / 5);
+                        // don't overwrite if higher priority present
+                        if (!costsDiv5ToUpgrade.has(processedCost)) {
```

Comment:
Ah ok cool I get the priority now. Can you leave a comment up where you use the term "deprioritize" specifying that it's used to break ties between matching-cost tower upgrades?